### PR TITLE
JDK-8285939: javadoc java.lang.Record should not have "Direct Known Subclasses:" section

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
As described in the JBS issue, the observed problem is a side-effect of a related but different issue, which is that record classes are not treated the same was as enum classes when it comes to generating the class hierarchy in `ClassTree`. Because record classes are not treated specially/differently, they are treated as ordinary/plain classes, with the side-effect that the page for `java.lang.Record` shows `Direct Known Subclasses`.

The underlying fix is therefore to clone the enum support in `ClassTree` and related classes, to provide support for record classes. It is possible to do an extreme minimal clone, but that just clones some of the messy evolution already there. Instead, the code in `ClassTree` and its clients is refactored and modernized.

Previously there were four explicit pairs of member fields, containing data for different groups (hierarchies) of classes, namely: plain classes, enum classes, interfaces and annotation interfaces. These fields are most of the code to support them are moved into some new abstractions to encapsulate related functionality.

1. The main new abstraction in `ClassTree` is `Hierarchy` which provides the support for the different hierarchies displayed in the generated pages.
2. A new enum `HierarchyKind` identifies the four existing hierarchies (listed above) and now a new one, for record classes. The hierarchies correspond to the different kinds of declared type.
3. A small new class `SubtypeMap` which is a multi-map for mapping a type element to its subtypes.  This is used in `Hierarchy` and to record the classes that implement an interfaces.

Generally, the naming should be clearer and more obvious. The most confusing name in the old code was `enumSubtypes` which was weird because enum classes don't have subtypes.  It meant "subtypes of supertypes of enum classes". This was a prime motivator to switch to the `hierarchy` terminology.   The variant spellings of `intfacs` have all been replaced  by `interfaces`, and `classtree` becomes `classTree`.

*Testing*: a new test case has been added to the `TestRecordTypes.java` test, which verifies the new record hierarchy is displayed on a a package tree page.  It is not simple to directly test the observed/reported behavior, because it is specific to the JDK API documentation, and because it is about the content of the `java.lang.Record` API page. However, manual inspection and diff comparison between JDK API documentation generated before and after this change reveals only expected differences. These differences are on the `java.lang.R4cord` page (where the offending section is no longer displayed) and on the pages related to the two existing records in JDK ... which are now listed in `Record Class Hierarchy` sections in the appropriate `package-tree.html` files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8285939](https://bugs.openjdk.java.net/browse/JDK-8285939): javadoc java.lang.Record should not have "Direct Known Subclasses:" section

### Reviewers
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to [a4d52be4](https://git.openjdk.java.net/jdk/pull/8523/files/a4d52be45855fb5a6453a76516c2621fecec31c4)
 * [Hannes Wallnöfer](https://openjdk.java.net/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to [c8ec372f](https://git.openjdk.java.net/jdk/pull/8523/files/c8ec372f89b060cc7061dc7a040d4dce5e9d9dab)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8523/head:pull/8523` \
`$ git checkout pull/8523`

Update a local copy of the PR: \
`$ git checkout pull/8523` \
`$ git pull https://git.openjdk.java.net/jdk pull/8523/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8523`

View PR using the GUI difftool: \
`$ git pr show -t 8523`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8523.diff">https://git.openjdk.java.net/jdk/pull/8523.diff</a>

</details>

### CODE REVIEW

**Review by pavelrappo — COMMENTED** [review]
Please merge this branch with the master (they are now in conflict) and I will review the change.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractTreeWriter.java [hunk 1]
```diff
@@ -71,56 +72,41 @@ protected AbstractTreeWriter(HtmlConfiguration configuration,
...
-     */
-    protected void addTree(SortedSet<TypeElement> sset, String heading, Content content) {
-        addTree(sset, heading, content, false);
-    }
-
-    protected void addTree(SortedSet<TypeElement> sset, String heading,
-                           Content content, boolean isEnums) {
-        if (!sset.isEmpty()) {
-            TypeElement firstTypeElement = sset.first();
+    protected void addTree(Hierarchy hierarchy, String heading, Content content) {
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractTreeWriter.java:94]: Noted; will do.

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/AbstractTreeWriter.java:94]: done

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java [hunk 2]
```diff
@@ -42,233 +41,264 @@
...
+
+        /**
+         * {@return the subtypes of the given type element, or an empty set if there are none}
+         * @param typeElement the type element
+         */
+        public SortedSet<TypeElement> subtypes(TypeElement typeElement) {
+            return subtypes.getOrDefault(typeElement, Collections.emptySortedSet());
+        }
+
+        public SortedSet<TypeElement> allSubtypes(TypeElement typeElement) {
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java:108]: done

File: test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java [hunk 3]
```diff
@@ -605,4 +605,41 @@ public void testSerializableType(Path base) throws IOException {
                     </ul>
                     </section>""");
     }
+
+    @Test
+    public void testPackageTree(Path base) throws IOException {
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java:610]: Am I right saying that we cannot easily test the original bug symptom, which is the presence of "Direct Known Subclasses:" on the api/java.base/java/lang/Record.html page?

**Review by pavelrappo — APPROVED** [review]
Looks good.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java [hunk 4]
```diff
@@ -42,233 +41,274 @@
...
+         * @param typeElement the type element
+         */
+        public SortedSet<TypeElement> subtypes(TypeElement typeElement) {
+            return subtypes.getOrDefault(typeElement, Collections.emptySortedSet());
+        }
+
+        /**
+         * {@return the set of all subtypes of the given type element, or an empty set if there are none}
+         *
+         * The set of all subtypes is the transitive closure of the {@linkplain #subtypes() immediate subtypes}
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java:103]: I had meant `#subtypes()` but now that that method has been deleted, `#subtypes(TypeElement)` will have to do.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java [hunk 5]
```diff
@@ -42,233 +41,274 @@
...
+
+        Hierarchy(Comparator<Element> comparator) {
+            this.comparator = comparator;
+            roots = new TreeSet<>(comparator);
+            subtypes = new SubtypeMap(comparator);
+        }
+
+        /**
+         * {@return the roots of the hierarchy}
+         * The roots are the classes or interfaces with no superclass or superinterfaces.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java:78]: OK

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java [hunk 6]
```diff
@@ -42,233 +41,274 @@
...
+            return out;
+        }
+    }
 
     /**
-     * List of base-interfaces. Contains set of all the interfaces who do not
-     * have super-interfaces. Can be used to get the mapped listing of
-     * sub-interfaces.
+     * A map giving the subtypes for each of a collection of type elements.
+     * The subtypes may be subclasses or subinterfaces, depending on the context.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java:128]: [JDK-8287606](https://bugs.openjdk.java.net/browse/JDK-8287606)

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java [hunk 7]
```diff
@@ -42,233 +41,274 @@
...
+
         buildTree(classesSet);
     }
 
     /**
-     * Generate mapping for the sub-classes for every class in this run.
-     * Return the sub-class set for java.lang.Object which will be having
-     * sub-class listing for itself and also for each sub-class itself will
-     * have their own sub-class lists.
+     * Generate the hierarchies for the given set of type elements.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/ClassTree.java:228]: Done

File: test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java (hunk 3 above)
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testRecordTypes/TestRecordTypes.java:610]: Yes, that is a specific test on a specific page in the JDK API documentation.

A possible test would involve ensuring access to the `src/` directory, and then maybe running javadoc on a subset of APIs, such as `java.lang.*` with maybe dummy placeholders for JDK custom tags, like `@jls`, `@jvms`, `@moduleGraph` etc.

**Pull request closed without merging**